src/hotspot/share/opto/vtransform.cpp

```diff
@@ -0,0 +1,441 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "opto/vtransform.hpp"
+#include "opto/vectornode.hpp"
+#include "opto/convertnode.hpp"
+
+void VTransformGraph::add_vtnode(VTransformNode* vtnode) {
+  assert(vtnode->_idx == _vtnodes.length(), "position must match idx");
+  _vtnodes.push(vtnode);
+}
+
+// Compute a linearization of the graph. We do this with a reverse-post-order of a DFS.
+// This only works if the graph is a directed acyclic graph (DAG). The C2 graph, and
+// the VLoopDependencyGraph are both DAGs, but after introduction of vectors/packs, the
+// graph has additional constraints which can introduce cycles. Example:
+//
+//                                                       +--------+
+//  A -> X                                               |        v
+//                     Pack [A,B] and [X,Y]             [A,B]    [X,Y]
+//  Y -> B                                                 ^        |
+//                                                         +--------+
+//
+// We return "true" IFF we find no cycle, i.e. if the linearization succeeds.
+bool VTransformGraph::schedule() {
```

Comment:
Note: used to be `PacksetGraph::schedule`. I use a reverse-post-order algorithm now, used to be topsort.